File: assisted_basic_mvn_project/src/main/java/com/learning/Factorial.java
package com.learning;

/**
 * The Factorial class provides utility methods for calculating factorial values.
 * A factorial of a non-negative integer n is the product of all positive integers
 * less than or equal to n. For example, 5! = 5 × 4 × 3 × 2 × 1 = 120.
 */
public class Factorial {
  
    /**
     * Calculates the factorial of a given non-negative integer.
     *
     * @param n the non-negative integer for which to calculate the factorial
     * @return the factorial of n as a long value
     * @throws IllegalArgumentException if n is negative
     * 
     * @example
     * factorial(5) returns 120
     * factorial(0) returns 1
     */
    public static long factorial(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("n must be non-negative");
        }
        long result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }

    /**
     * Main method that demonstrates the usage of the factorial calculation.
     * Computes and prints the factorial of the number 5.
     *
     * @param args command line arguments (not used)
     */
    public static void main(String[] args) {
        int number = 5;
        System.out.println("Factorial of " + number + " is: " + factorial(number));
    }
}

File: assisted_basic_python_project/factorial.py
# Calculate factorial of a number
# Factorial of n (denoted as n!) is the product of all positive integers less than or equal to n
# Example: 5! = 5 × 4 × 3 × 2 × 1 = 120

def factorial(n):
    """
    Calculate the factorial of a non-negative integer n.
    
    Args:
        n: A non-negative integer
        
    Returns:
        The factorial of n
        
    Raises:
        ValueError: If n is negative or not an integer
    """
    if not isinstance(n, int):
        raise ValueError("Input must be an integer")
    
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    
    if n == 0 or n == 1:
        return 1
    
    result = 1
    for i in range(2, n + 1):
        result *= i
    
    return result


# Test cases
if __name__ == "__main__":
    print(f"factorial(0) = {factorial(0)}")    # 1
    print(f"factorial(1) = {factorial(1)}")    # 1
    print(f"factorial(5) = {factorial(5)}")    # 120
    print(f"factorial(10) = {factorial(10)}")  # 3628800
